### PR TITLE
[feature](nereids)support one expression-rewrite rule: inPredicateToEqualTo

### PR DESCRIPTION
**Pull request by YangShaw**
# Proposed changes

Issue Number: close #xxx

## Problem summary
Add one expression rewrite rule:
rewrite InPredicate to an EqualTo Expression, if there exists exactly one element in InPredicate Options.

Examples:
 * where A in (x) ==> where A = x
 * where A not in (x) ==> where not A = x

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * where A in (x) ==> where A = x
+ * where A not in (x) ==> where not A = x (After ExpressionTranslator, "not A = x" will be translated to "A != x")
+ */
+public class InPredicateToEqualToRule extends AbstractExpressionRewriteRule {
+
+    public static InPredicateToEqualToRule INSTANCE = new InPredicateToEqualToRule();
+
+    @Override
+    public Expression visitInPredicate(InPredicate inPredicate, ExpressionRewriteContext context) {
+        if (inPredicate.getOptions().size() > 1) {
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java:39]:
```suggestion
        if (inPredicate.getOptions().size() != 1) {
```

File: regression-test/suites/nereids_syntax_p0/inpredicate.groovy [hunk 2]
```diff
@@ -55,5 +55,13 @@ suite("inpredicate") {
     order_qt_in_predicate_8 """
         SELECT * FROM supplier WHERE s_nation not in ('PERU', 'ETHIOPIA');
     """
+
+    order_qt_in_predicate_9 """
+        SELECT * FROM supplier WHERE s_suppkey in (1);
```
Comment (924060929) [inline regression-test/suites/nereids_syntax_p0/inpredicate.groovy:60]:
```suggestion
        SELECT * FROM supplier WHERE s_suppkey in (15);
```

File: regression-test/suites/nereids_syntax_p0/inpredicate.groovy [hunk 3]
```diff
@@ -55,5 +55,13 @@ suite("inpredicate") {
...
+    """
+
+    order_qt_in_predicate_10 """
+        SELECT * FROM supplier WHERE s_suppkey not in (1);
```
Comment (924060929) [inline regression-test/suites/nereids_syntax_p0/inpredicate.groovy:64]:
```suggestion
        SELECT * FROM supplier WHERE s_suppkey not in (15);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java (hunk 1 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            return inPredicate;
+        }
+
+        Expression left = inPredicate.getCompareExpr();
+        Expression right = inPredicate.getOptions().get(0);
+
+        if (right instanceof InPredicate) {
+            right = simplifyInnerIn((InPredicate) right);
+        }
+        return new EqualTo(left, right);
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java:53]:
```suggestion
        return new EqualTo(left, right.accept(this, context));
```

**Comment by wangshuo128** [comment]
I think this patch should be a feature rather than an enhancement. :)

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java (hunk 4 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/InPredicateToEqualToRule.java:53]: Wow! Amazing~ But I still don't understand why the left expr need to call accept() also

**Comment by YangShaw** [comment]
>

**Comment by YangShaw** [comment]
> I think this patch should be a feature rather than an enhancement. :)

Ok, I will edit

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java [hunk 5]
```diff
@@ -151,6 +152,19 @@ public void testBetweenToCompoundRule() {
...
+    @Test
+    public void testInPredicateToEqualToRule() {
+        executor = new ExpressionRuleExecutor(ImmutableList.of(InPredicateToEqualToRule.INSTANCE));
+
+        assertRewrite("a in (1)", "a = 1");
+        assertRewrite("a in (1, 2)", "a in (1, 2)");
+        assertRewrite("a not in (1)", "not a = 1");
+        assertRewrite("a not in (1, 2)", "not a in (1, 2)");
+        assertRewrite("a in (a in (1))", "a = (a = 1)");
+        assertRewrite("a in (a in (1, 2))", "a = (a in (1, 2))");
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java:164]: add a left case, such as
```sql
case a when b in (1) then a else c end in (1)
```

Comment (YangShaw) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/rules/expression/rewrite/ExpressionRewriteTest.java:164]: done

**Review by 924060929 — APPROVED** [review]
LGTM

**Comment by YangShaw** [comment]
Still have little bug about recursive, like:
``` a in (1, 2, (a in (1)))```
This kind of cases will be resolved in next pr.

**Pull request merged**